File: backend/handlers/screener.go
package handlers

import (
	"log"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	"investorcenter-api/database"
	"investorcenter-api/models"

	"github.com/gin-gonic/gin"
)

const (
	// defaultScreenerLimit is the page size when no limit is specified.
	defaultScreenerLimit = 100
	// maxScreenerLimit caps the maximum rows a single request can return.
	maxScreenerLimit = 20000
)

// GetScreenerStocks handles the stock screener endpoint
// GET /api/v1/screener/stocks
func GetScreenerStocks(c *gin.Context) {
	// Check database connection
	if database.DB == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"error":   "Database not available",
			"message": "Screener service is temporarily unavailable",
		})
		return
	}

	// Parse query parameters
	params := parseScreenerParams(c)

	// Fetch stocks from database
	stocks, total, err := database.GetScreenerStocks(params)
	if err != nil {
		log.Printf("Error fetching screener stocks: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to fetch stocks",
			"message": "An error occurred while retrieving screener data",
		})
		return
	}

	// Calculate total pages
	totalPages := int(math.Ceil(float64(total) / float64(params.Limit)))

	// Build response
	response := models.ScreenerResponse{
		Data: stocks,
		Meta: models.ScreenerMeta{
			Total:      total,
			Page:       params.Page,
			Limit:      params.Limit,
			TotalPages: totalPages,
			Timestamp:  time.Now().UTC().Format(time.RFC3339),
		},
	}

	c.JSON(http.StatusOK, response)
}

// floatParam is a declarative definition for parsing a float64 query parameter.
type floatParam struct {
	key    string                                           // URL query key
	setter func(params *models.ScreenerParams, val float64) // Sets the value on ScreenerParams
}

// rangeParams defines all min/max float query parameters supported by the screener.
// Adding a new range filter requires:
//  1. Add fields to ScreenerParams (models/stock.go)
//  2. Add entry to RangeFilters (database/filter_registry.go)
//  3. Add entries here for URL param parsing
var rangeParams = []floatParam{
	// Market data
	{key: "market_cap_min", setter: func(p *models.ScreenerParams, v float64) { p.MarketCapMin = &v }},
	{key: "market_cap_max", setter: func(p *models.ScreenerParams, v float64) { p.MarketCapMax = &v }},

	// Valuation
	{key: "pe_min", setter: func(p *models.ScreenerParams, v float64) { p.PEMin = &v }},
	{key: "pe_max", setter: func(p *models.ScreenerParams, v float64) { p.PEMax = &v }},
	{key: "pb_min", setter: func(p *models.ScreenerParams, v float64) { p.PBMin = &v }},
	{key: "pb_max", setter: func(p *models.ScreenerParams, v float64) { p.PBMax = &v }},
	{key: "ps_min", setter: func(p *models.ScreenerParams, v float64) { p.PSMin = &v }},
	{key: "ps_max", setter: func(p *models.ScreenerParams, v float64) { p.PSMax = &v }},

	// Profitability
	{key: "roe_min", setter: func(p *models.ScreenerParams, v float64) { p.ROEMin = &v }},
	{key: "roe_max", setter: func(p *models.ScreenerParams, v float64) { p.ROEMax = &v }},
	{key: "roa_min", setter: func(p *models.ScreenerParams, v float64) { p.ROAMin = &v }},
	{key: "roa_max", setter: func(p *models.ScreenerParams, v float64) { p.ROAMax = &v }},
	{key: "gross_margin_min", setter: func(p *models.ScreenerParams, v float64) { p.GrossMarginMin = &v }},
	{key: "gross_margin_max", setter: func(p *models.ScreenerParams, v float64) { p.GrossMarginMax = &v }},
	{key: "net_margin_min", setter: func(p *models.ScreenerParams, v float64) { p.NetMarginMin = &v }},
	{key: "net_margin_max", setter: func(p *models.ScreenerParams, v float64) { p.NetMarginMax = &v }},

	// Financial health
	{key: "de_min", setter: func(p *models.ScreenerParams, v float64) { p.DebtToEquityMin = &v }},
	{key: "de_max", setter: func(p *models.ScreenerParams, v float64) { p.DebtToEquityMax = &v }},
	{key: "current_ratio_min", setter: func(p *models.ScreenerParams, v float64) { p.CurrentRatioMin = &v }},
	{key: "current_ratio_max", setter: func(p *models.ScreenerParams, v float64) { p.CurrentRatioMax = &v }},

	// Growth
	{key: "revenue_growth_min", setter: func(p *models.ScreenerParams, v float64) { p.RevenueGrowthMin = &v }},
	{key: "revenue_growth_max", setter: func(p *models.ScreenerParams, v float64) { p.RevenueGrowthMax = &v }},
	{key: "eps_growth_min", setter: func(p *models.ScreenerParams, v float64) { p.EPSGrowthMin = &v }},
	{key: "eps_growth_max", setter: func(p *models.ScreenerParams, v float64) { p.EPSGrowthMax = &v }},

	// Dividends
	{key: "dividend_yield_min", setter: func(p *models.ScreenerParams, v float64) { p.DividendYieldMin = &v }},
	{key: "dividend_yield_max", setter: func(p *models.ScreenerParams, v float64) { p.DividendYieldMax = &v }},
	{key: "payout_ratio_min", setter: func(p *models.ScreenerParams, v float64) { p.PayoutRatioMin = &v }},
	{key: "payout_ratio_max", setter: func(p *models.ScreenerParams, v float64) { p.PayoutRatioMax = &v }},
	{key: "consec_div_years_min", setter: func(p *models.ScreenerParams, v float64) { p.ConsecutiveDivYearsMin = &v }},

	// Risk
	{key: "beta_min", setter: func(p *models.ScreenerParams, v float64) { p.BetaMin = &v }},
	{key: "beta_max", setter: func(p *models.ScreenerParams, v float64) { p.BetaMax = &v }},

	// Fair value
	{key: "dcf_upside_min", setter: func(p *models.ScreenerParams, v float64) { p.DCFUpsideMin = &v }},
	{key: "dcf_upside_max", setter: func(p *models.ScreenerParams, v float64) { p.DCFUpsideMax = &v }},

	// IC Score
	{key: "ic_score_min", setter: func(p *models.ScreenerParams, v float64) { p.ICScoreMin = &v }},
	{key: "ic_score_max", setter: func(p *models.ScreenerParams, v float64) { p.ICScoreMax = &v }},

	// IC Score sub-factors
	{key: "value_score_min", setter: func(p *models.ScreenerParams, v float64) { p.ValueScoreMin = &v }},
	{key: "value_score_max", setter: func(p *models.ScreenerParams, v float64) { p.ValueScoreMax = &v }},
	{key: "growth_score_min", setter: func(p *models.ScreenerParams, v float64) { p.GrowthScoreMin = &v }},
	{key: "growth_score_max", setter: func(p *models.ScreenerParams, v float64) { p.GrowthScoreMax = &v }},
	{key: "profitability_score_min", setter: func(p *models.ScreenerParams, v float64) { p.ProfitabilityScoreMin = &v }},
	{key: "profitability_score_max", setter: func(p *models.ScreenerParams, v float64) { p.ProfitabilityScoreMax = &v }},
	{key: "financial_health_score_min", setter: func(p *models.ScreenerParams, v float64) { p.FinancialHealthScoreMin = &v }},
	{key: "financial_health_score_max", setter: func(p *models.ScreenerParams, v float64) { p.FinancialHealthScoreMax = &v }},
	{key: "momentum_score_min", setter: func(p *models.ScreenerParams, v float64) { p.MomentumScoreMin = &v }},
	{key: "momentum_score_max", setter: func(p *models.ScreenerParams, v float64) { p.MomentumScoreMax = &v }},
	{key: "analyst_score_min", setter: func(p *models.ScreenerParams, v float64) { p.AnalystScoreMin = &v }},
	{key: "analyst_score_max", setter: func(p *models.ScreenerParams, v float64) { p.AnalystScoreMax = &v }},
	{key: "insider_score_min", setter: func(p *models.ScreenerParams, v float64) { p.InsiderScoreMin = &v }},
	{key: "insider_score_max", setter: func(p *models.ScreenerParams, v float64) { p.InsiderScoreMax = &v }},
	{key: "institutional_score_min", setter: func(p *models.ScreenerParams, v float64) { p.InstitutionalScoreMin = &v }},
	{key: "institutional_score_max", setter: func(p *models.ScreenerParams, v float64) { p.InstitutionalScoreMax = &v }},
	{key: "sentiment_score_min", setter: func(p *models.ScreenerParams, v float64) { p.SentimentScoreMin = &v }},
	{key: "sentiment_score_max", setter: func(p *models.ScreenerParams, v float64) { p.SentimentScoreMax = &v }},
	{key: "technical_score_min", setter: func(p *models.ScreenerParams, v float64) { p.TechnicalScoreMin = &v }},
	{key: "technical_score_max", setter: func(p *models.ScreenerParams, v float64) { p.TechnicalScoreMax = &v }},
}

// parseScreenerParams extracts and validates query parameters
func parseScreenerParams(c *gin.Context) models.ScreenerParams {
	params := models.ScreenerParams{
		Page:  1,
		Limit: defaultScreenerLimit,
		Sort:  "market_cap",
		Order: "desc",
		// AssetType is parsed below but not used in the query — the
		// screener_data materialized view already filters to asset_type='CS'.
		// Kept for future use if the screener expands to ETFs.
		AssetType: "CS",
	}

	// Page
	if page, err := strconv.Atoi(c.DefaultQuery("page", "1")); err == nil && page > 0 {
		params.Page = page
	}

	// Limit (capped at maxScreenerLimit)
	if raw := c.Query("limit"); raw != "" {
		if limit, err := strconv.Atoi(raw); err == nil && limit > 0 {
			if limit > maxScreenerLimit {
				limit = maxScreenerLimit
			}
			params.Limit = limit
		}
	}

	// Sort field
	sort := c.DefaultQuery("sort", "market_cap")
	if _, ok := database.ValidScreenerSortColumns[sort]; ok {
		params.Sort = sort
	}

	// Sort order
	order := strings.ToLower(c.DefaultQuery("order", "desc"))
	if order == "asc" || order == "desc" {
		params.Order = order
	}

	// Sectors (comma-separated)
	if sectors := c.Query("sectors"); sectors != "" {
		params.Sectors = strings.Split(sectors, ",")
		for i := range params.Sectors {
			params.Sectors[i] = strings.TrimSpace(params.Sectors[i])
		}
	}

	// Industries (comma-separated)
	if industries := c.Query("industries"); industries != "" {
		params.Industries = strings.Split(industries, ",")
		for i := range params.Industries {
			params.Industries[i] = strings.TrimSpace(params.Industries[i])
		}
	}

	// Parse all range filters declaratively
	for _, fp := range rangeParams {
		if raw := c.Query(fp.key); raw != "" {
			if val, err := strconv.ParseFloat(raw, 64); err == nil {
				fp.setter(&params, val)
			}
		}
	}

	// Asset type (validated against allowlist)
	if assetType := c.Query("asset_type"); assetType != "" {
		validAssetTypes := map[string]bool{
			"CS": true, "ETF": true, "ADRC": true, "ADRW": true,
			"WARRANT": true, "RIGHT": true, "UNIT": true,
			"PFD": true, "FUND": true, "SP": true, "OS": true,
			"crypto": true,
		}
		if validAssetTypes[assetType] {
			params.AssetType = assetType
		}
		// Invalid values silently fall back to default "CS"
	}

	return params
}

File: lib/types/screener.ts
// Screener types — aligned with backend models/stock.go ScreenerStock + ScreenerParams

/** A single stock row returned by the screener API */
export interface ScreenerStock {
  // Core identity
  symbol: string;
  name: string;
  sector: string | null;
  industry: string | null;

  // Market data
  market_cap: number | null;
  price: number | null;

  // Valuation
  pe_ratio: number | null;
  pb_ratio: number | null;
  ps_ratio: number | null;

  // Profitability
  roe: number | null;
  roa: number | null;
  gross_margin: number | null;
  operating_margin: number | null;
  net_margin: number | null;

  // Financial health
  debt_to_equity: number | null;
  current_ratio: number | null;

  // Growth
  revenue_growth: number | null;
  eps_growth_yoy: number | null;

  // Dividends
  dividend_yield: number | null;
  payout_ratio: number | null;
  consecutive_dividend_years: number | null;

  // Risk
  beta: number | null;

  // Fair value
  dcf_upside_percent: number | null;

  // IC Score
  ic_score: number | null;
  ic_rating: string | null;

  // IC Score sub-factors (0-100)
  value_score: number | null;
  growth_score: number | null;
  profitability_score: number | null;
  financial_health_score: number | null;
  momentum_score: number | null;
  analyst_consensus_score: number | null;
  insider_activity_score: number | null;
  institutional_score: number | null;
  news_sentiment_score: number | null;
  technical_score: number | null;

  // IC Score metadata
  ic_sector_percentile: number | null;
  lifecycle_stage: string | null;
}

/** Pagination metadata from the screener API */
export interface ScreenerMeta {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  timestamp: string;
}

/** Full API response shape */
export interface ScreenerResponse {
  data: ScreenerStock[];
  meta: ScreenerMeta;
}

/** Query parameters sent to the screener API. All optional except page/limit/sort/order. */
export interface ScreenerApiParams {
  // Pagination & sorting
  page?: number;
  limit?: number;
  sort?: string;
  order?: 'asc' | 'desc';

  // Categorical
  sectors?: string;
  industries?: string;
  asset_type?: string;

  // Market data
  market_cap_min?: number;
  market_cap_max?: number;

  // Valuation
  pe_min?: number;
  pe_max?: number;
  pb_min?: number;
  pb_max?: number;
  ps_min?: number;
  ps_max?: number;

  // Profitability
  roe_min?: number;
  roe_max?: number;
  roa_min?: number;
  roa_max?: number;
  gross_margin_min?: number;
  gross_margin_max?: number;
  net_margin_min?: number;
  net_margin_max?: number;

  // Financial health
  de_min?: number;
  de_max?: number;
  current_ratio_min?: number;
  current_ratio_max?: number;

  // Growth
  revenue_growth_min?: number;
  revenue_growth_max?: number;
  eps_growth_min?: number;
  eps_growth_max?: number;

  // Dividends
  dividend_yield_min?: number;
  dividend_yield_max?: number;
  payout_ratio_min?: number;
  payout_ratio_max?: number;
  consec_div_years_min?: number;

  // Risk
  beta_min?: number;
  beta_max?: number;

  // Fair value
  dcf_upside_min?: number;
  dcf_upside_max?: number;

  // IC Score
  ic_score_min?: number;
  ic_score_max?: number;

  // IC Score sub-factors
  value_score_min?: number;
  value_score_max?: number;
  growth_score_min?: number;
  growth_score_max?: number;
  profitability_score_min?: number;
  profitability_score_max?: number;
  financial_health_score_min?: number;
  financial_health_score_max?: number;
  momentum_score_min?: number;
  momentum_score_max?: number;
  analyst_score_min?: number;
  analyst_score_max?: number;
  insider_score_min?: number;
  insider_score_max?: number;
  institutional_score_min?: number;
  institutional_score_max?: number;
  sentiment_score_min?: number;
  sentiment_score_max?: number;
  technical_score_min?: number;
  technical_score_max?: number;
}

/** Valid sort columns — same keys as backend ValidScreenerSortColumns */
export type ScreenerSortField =
  | 'symbol'
  | 'name'
  | 'market_cap'
  | 'price'
  | 'pe_ratio'
  | 'pb_ratio'
  | 'ps_ratio'
  | 'roe'
  | 'roa'
  | 'gross_margin'
  | 'operating_margin'
  | 'net_margin'
  | 'debt_to_equity'
  | 'current_ratio'
  | 'revenue_growth'
  | 'eps_growth_yoy'
  | 'dividend_yield'
  | 'payout_ratio'
  | 'consecutive_dividend_years'
  | 'beta'
  | 'dcf_upside_percent'
  | 'ic_score'
  | 'value_score'
  | 'growth_score'
  | 'profitability_score'
  | 'financial_health_score'
  | 'momentum_score'
  | 'analyst_consensus_score'
  | 'insider_activity_score'
  | 'institutional_score'
  | 'news_sentiment_score'
  | 'technical_score';

/** Preset screen definition */
export interface ScreenerPreset {
  id: string;
  name: string;
  description: string;
  params: Partial<ScreenerApiParams>;
}
